File: 113-path-sum-ii/113-path-sum-ii.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
import scala.collection.mutable._

object Solution {
    def pathSum(root: TreeNode, targetSum: Int): List[List[Int]] = {
        val res: ListBuffer[List[Int]] = ListBuffer()

        helper(root, targetSum, Stack(), res)

        res.toList
    }

    private def helper(root: TreeNode, targetSum: Int, current: Stack[Int], list: ListBuffer[List[Int]]): Unit = {
        if (root != null) {
            if (root.value == targetSum && root.left == null && root.right == null) {
                current.push(root.value)
                list += current.toList.reverse
                current.pop
            } else {
                val sum = targetSum - root.value

                current.push(root.value)

                helper(root.left, sum, current, list)
                helper(root.right, sum, current, list)

                current.pop
            }
        }
    }
}

File: 113-path-sum-ii/113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        return dfs(root, targetSum);
    }
    
    List<List<Integer>> dfs(TreeNode root, int targetSum) {
        List<List<Integer>> result = new ArrayList<>();
        
        if (root != null) dfsHelper(root, targetSum, new ArrayList<>(), result);
        
        return result;
    }
    
    void dfsHelper(TreeNode root, int sum, List<Integer> current, List<List<Integer>> result) {
        current.add(root.val);
        
        if (root.left == null && root.right == null) {
            if (root.val == sum) result.add(List.copyOf(current));
            current.remove(current.size() - 1);
            return;
        }
        
        if (root.left != null) dfsHelper(root.left, sum - root.val, current, result);
        if (root.right != null) dfsHelper(root.right, sum - root.val, current, result);
        
        current.remove(current.size() - 1);
    }
}